lsi/rom/tools/cp16mic.py

```diff
@@ -0,0 +1,981 @@
+    #
+    # Calculates the expression in Polish Reverse Notation
+    #
+    def calc_exp(self, polish):
+        stack = []
+        for token in polish:
+            if token in wd16.EXP_OPERS:
+                if token[0] == 'U':
+                    x = stack.pop()
+                    stack.append(wd16.EXP_OPERS[token][1](x))
+                else:
+                    y, x = stack.pop(), stack.pop()
+                    stack.append(wd16.EXP_OPERS[token][1](x, y))
+            else:
+                stack.append(token)
+        if not stack:
+            raise SyntaxError("invalid integer expression (empty)")
+        return stack[0]
+    #
+    # Converts the infix expression to Polish Reverse Notation
+    #
+    def polish_exp(self, string):
+        stack = []
+        for token in string:
+            if token in wd16.EXP_OPERS:
+                while (stack and stack[-1] != "(" and
+                    wd16.EXP_OPERS[token][0] <= wd16.EXP_OPERS[stack[-1]][0]):
+                    yield stack.pop()
+                stack.append(token)
+            elif token == ")":
+                while stack:
+                    x = stack.pop()
+                    if x == "(":
+                        break
+                    yield x
+            elif token == "(":
+                stack.append(token)
+            else:
+                yield token
+        while stack:
+            yield stack.pop()
+    #
+    # Token generator wrapper, replaces [+-~] with unary tags
+    #
+    def subst_unary(self, string):
+        prev = True
+        for token in string:
+            if type(token) is not str:
+                prev = False
+                yield token
+                continue
+            if token in wd16.EXP_UNA:
+                if prev:
+                    yield 'U' + token
+            else:
+                prev = token in wd16.EXP_OPS or token == '('
+                yield token
+    #
+    # Token generator wrapper, replaces the numbers with actual integers
+    #
+    def subst_number(self, token):
+        try:
+            lt = len(token)
+            if token[0] == '0':
+                if lt > 1:
+                    if token[1] in 'xX':
+                        return int(token[2:], base=16)
+                    if token[1] in 'bB':
+                        return int(token[2:], base=2)
+                    if token[-1] == '.':
+                        return int(token[:-1], base=10)
+                    return int(token[:-1], base=8)
+            if token[-1] == '.':
+                return int(token[:-1], base=10)
+            return int(token, self.radx)
+        except ValueError:
+            raise SyntaxError("invalid integer expression '%s'" % token)
+    #
+    # Token generator wrapper, replaces the symbols with actual integers
+    #
+    def subst_symbol(self, string):
+        for token in string:
+            if token in wd16.EXP_ALL:
+                yield token
+            elif token[0] in '0123456789':
```

Comment:
do you take into account value of `self.radx` == 8 or 10 there?

Reply:
BTW, `subst_number()` supports hexadecimal constants

Reply:
Valid identifier can't start with digits, so it is filtering out the potential digits. If radix is 8 then int() will raise exception later

Reply:
> BTW, `subst_number()` supports hexadecimal constants
Yes:
- 0xnnn is always treated as hex
- 0bnnn  is always treated as bin
- nnnn. is always treated as dec
- 0nnn is always treated as oct
- nnnn is treated on radix settings